### ISSUE TITLE
Usage of comms for messages without explicit size fields

### ISSUE THREAD

**Issue by melg8** [commschamp/comms_champion#9]
Hi, I really like your work  with comms and great documentation about how to use it and how it works. I want to use it for handling messages in embedded environment, but I've encountered one problem. Your library explicitly state MsgSizeLayer which obtains data about size of message from some values inside the message.
Now, let's talk about messages which I have to send/receive:
They are something similar to ByteStuffed messages from [this](https://web.cs.wpi.edu/~rek/Undergrad_Nets/C02/BitByteStuffing.pdf) slide 9.
So I have several differences in messages:
1. There is no explicit size known at the start of the message, the message is finished if some suffix bytes are encountered in byte stream.
2. Message suffix consist of two bytes, and if in original payload we encounter byte with same start value it should be stuffed - changed with two bytes, for example if suffix is 0xFF 0xFE  we must change all bytes in payload from 0xFF to 0xFF 0xFA - so we can differ - is this really end of message, or this is only single byte from payload. 
3. Because of this bytestuffing - in the worst case - obtained message can be x2 the size of the original, and I can't store the whole message before unstuffing it - I should unstuff elements and store them "on fly" - when message bytes arriving.
4. Message couldn't be obtained in single piece it come in chunks which are 2-8 times smaller than the whole thing.
5. When sending message - byte stuffing and suffixing must be applied as well.
6. Message have prefix which can differ in second byte, for example 0xFF 0x01 and 0xFF 0x02 which mean different "type" of message - is it incoming command or is it answer to my command.
7. This is not COBS, I can't use it, or change protocol to have explicit size field.

Other than that - I have as usual id, sync, payload, and crc  elements inside the message. 

I've encounter messaging task for the first time, really new to the field of messaging, and doesn't want to handle the problem from scratch, so I've tried to find some building blocks solutions for that problem and encountered your library, which looks very promising to me, but I'm worried if it is possible to handle such types of messages using your library, or it doesn't fit well in terms of your library. 

Could you estimate what should be done to adapt your library for handling such byte stuffed/size-less messages and which steps it will require?
I would really appreciate any help.

**Comment by arobenko** [comment]
Hi @melg8,
The COMMS library is implemented in a way that allows you to write any complimentary logic yourself in case such one is not provided "out of the box". Nothing really prevents you from doing extra pre- and post- processing of the input / output data. I would recommend using a pair of input and a pair of output buffers. Let's call them:
- **Buffer1** - contains raw data received from (or about to be sent over) the I/O link.
- **Buffer2** - contains single message with all the stuffing removed.

All the new incoming data you can put into the **Buffer1**. Then you do some pre-processing to determine the end of the message. Once the latter is detected you copy the discovered message data (while removing stuffing in the process) to the **Buffer2**. After that you can process the data using the framing defined using the COMMS library.

When sending the message the operation is reversed: you write message into the output **Buffer2** in the regular way, then you do post-processing of the written data by copying it to the output **Buffer1** adding stuffing and putting the required suffix at the end. After the post-processing, the data in the **Buffer1** is ready to be sent over the I/O link.

In the described above logic you can define your framing to be `SYNC | ID | PAYLOAD | CHECKSUM` without any size information. When invoking the read operation on the frame you just provide the size of the data inside the **Buffer2** (which always contains single message at a time).

Also I strongly recommend using the [commsdsl2comms](https://github.com/commschamp/commsdsl) code generator instead of trying to define the protocol manually. It will save you a lot of time.

The bottom line, I do not really see a need in any additional functionality required for the library itself.

Hope it helps.

P.S. I suppose you'll have other questions when you start defining your protocol. I suggest you write a direct [e-mail](https://github.com/commschamp/comms_champion#contact-information) to me with your questions and/or clarifications unless you want to keep this information public.

Regards,
Alex

**Comment by melg8** [comment]
Thanks for really quick answer! I'll try suggested approach. If there will be some follow-up question I'll send you a message directly.
